### PR TITLE
Eric/competition

### PR DESCRIPTION
**Pull request by condnsdmatters**
This is the branch that contains the Challenge Environment and the NeurIPS baseline we will release for it. 

The key differences from `NetHackChallenge` to `NetHackScore` are:
* All menus allowed
* Full key board action space allowed (with underlying code for quit, save and options switched off)
* Rotating characters (`characeter = '@'`)
*  `PREV_MSG` and `HISTORY` actions are removed (they seemed to be causing crashes)

The model itself we submit as baseline is based heavily on the model in Neurips2020 paper, and posted on the branch here (`neurips2020release`).  Here we have taken the best model and cut out all the other elements of the model, to keep the baseline simple, and to try to keep the model to only one file. 

The results of this run (for rotating characters and single characters are visible [here](https://fairwandb.org/nethack/competition3/reports/Starter-Kit--VmlldzoxMzk5) ). 

My suspicion is the model has gotten a bit slower (6000 sps) so there is probably room for a speed up. We can do this after the competition has started.

### CODE REVIEW

File: src/cmd.c [hunk 1]
```diff
@@ -3351,6 +3351,26 @@ int final;
     en_win = WIN_ERR;
 }
 
+int nle_dosave() {
+    pline("You get the feeling there's only one way to save yourself..."); 
+    return 1;
+}
+
+int nle_done2() {
+    pline("You can't quit now, you're having so much fun!");
```
Comment (tscmoo) [inline src/cmd.c:3360]: I love these messages

File: nle/agent/core/vtrace.py [hunk 2]
```diff
@@ -125,10 +125,7 @@ def from_importance_weights(
         vs = torch.add(vs_minus_v_xs, values)
 
         # Advantage for policy gradient.
-        broadcasted_bootstrap_values = torch.ones_like(vs[0]) * bootstrap_value
-        vs_t_plus_1 = torch.cat(
-            [vs[1:], broadcasted_bootstrap_values.unsqueeze(0)], dim=0
-        )
+        vs_t_plus_1 = torch.cat([vs[1:], torch.unsqueeze(bootstrap_value, 0)], dim=0)
```
Comment (tscmoo) [inline nle/agent/core/vtrace.py:128]: What is the rationale for changing this file? Are you sure the broadcasting isn't needed?
Just wondering as this file is not original, but copied from elsewhere.

File: nle/nethack/actions.py [hunk 3]
```diff
@@ -76,6 +89,12 @@ class MiscAction(enum.IntEnum):
     MORE = ord("\r")  # read the next message
 
 
+class UnsafeActions(enum.IntEnum):
+    # currently these result in an error or undesirable behaviour
+    HELP = ord("?")  # give a help message
+    PREVMSG = C("p")  # view recent game messages
```
Comment (tscmoo) [inline nle/nethack/actions.py:95]: So, is 'p' not in the action list?
Can this not be handled like quit, save, etc. below?

File: nle/scripts/play.py [hunk 4]
```diff
@@ -100,10 +101,14 @@ def play(env, mode, ngames, max_steps, seeds, savedir, no_render, render_mode, d
...
-                    print("Previous action: %s" % repr(env._actions[action]))
-                env.render(render_mode)
+                # print('--------')
+                # print(f"Previous reward: {str(reward):64s}")
+                act_str = repr(env._actions[action]) if action is not None else ''
+                # print(f"Previous action: {str(act_str):64s}" )
+                # print('--------')
+                # env.render(render_mode)
+                # print('--------')
+                # print('\033[31A') # Go up 31 lines.
```
Comment (tscmoo) [inline nle/scripts/play.py:111]: Are these lines supposed to be commented out? (also the ones below)

File: nle/nethack/actions.py (hunk 3 above)
Comment (condnsdmatters) [reply nle/nethack/actions.py:95]: No - control+P isnt. Luckily we dont need this action at all, so thats fine.

File: nle/scripts/play.py (hunk 4 above)
Comment (condnsdmatters) [reply nle/scripts/play.py:111]: will adjust these

File: nle/agent/core/vtrace.py (hunk 2 above)
Comment (condnsdmatters) [reply nle/agent/core/vtrace.py:128]: https://github.com/facebookresearch/nle/blob/0424d298f52f21e4e3f7dda09381051ce8359e30/nle/agent/core/vtrace.py#L128

Pulled straight out of our `neurips2020release` branch

File: nle/scripts/play.py (hunk 4 above)
Comment (condnsdmatters) [reply nle/scripts/play.py]: done

**Comment by condnsdmatters** [comment]
Although our CI is still down (and should be put up again asap) can confirm the code is all blacked, tests all pass, and the only flake8 errors we have are line too long in baseline model... (aren't these fine?)

**Comment by condnsdmatters** [comment]
Changes:
* Quitting has been reenabled, to allow for safe clean up of NLE
* The competition environment now aborts if you go 10000 steps without advancing in game timer
* You can load the polyhydra in test mode (it was broken before)

Needless to say I could squash restructure these commits, and maybe should for quick reversion.

File: nle/agent/core/vtrace.py (hunk 2 above)
Comment (heiner) [reply nle/agent/core/vtrace.py:128]: Let's leave this file as-is.

File: nle/env/base.py [hunk 5]
```diff
@@ -149,6 +149,22 @@
     "\x1b[95m",
     "\x1b[96m",
     "\x1b[97m",
+    "\033[30m",
```
Comment (heiner) [inline nle/env/base.py:152]: What is this? Why do we have this constant?

File: nle/env/base.py [hunk 6]
```diff
@@ -379,15 +403,18 @@ def step(self, action: int):
         last_observation = tuple(a.copy() for a in self.last_observation)
 
         observation, done = self.env.step(self._actions[action])
-        observation, done = self._perform_known_steps(
-            observation, done, exceptions=True
-        )
+        if (
+            not self._allow_all_modes or observation[self._program_state_index][0] == 1
```
Comment (heiner) [inline nle/env/base.py:407]: What does this do? Doesn't seem very readable to me.

File: nle/env/base.py [hunk 7]
```diff
@@ -379,15 +403,18 @@ def step(self, action: int):
...
+            observation, done = self._perform_known_steps(
+                observation, done, exceptions=True
+            )
 
         self._steps += 1
 
         self.last_observation = observation
 
-        if self._steps >= self._max_episode_steps:
+        if self._check_abort(observation):
```
Comment (heiner) [inline nle/env/base.py:417]: Are we going to override _check_abort in subclasses?

Comment (heiner) [reply nle/env/base.py:417]: Ah yes we do. Alright then.

File: nle/scripts/play.py [hunk 8]
```diff
@@ -100,10 +112,15 @@ def play(env, mode, ngames, max_steps, seeds, savedir, no_render, render_mode, d
...
-                if action is not None:
-                    print("Previous action: %s" % repr(env._actions[action]))
+                print("--------")
+                print(f"Previous reward: {str(reward):64s}")
+                act_str = repr(env._actions[action]) if action is not None else ""
+                print(f"Previous action: {str(act_str):64s}")
+                print("--------")
                 env.render(render_mode)
+                print("--------")
+                if not print_all_frames:
```
Comment (heiner) [inline nle/scripts/play.py:122]: That's a bit of a weird name for what it does (going back to override previous frames)?

File: nle/scripts/ttyplay.py [hunk 9]
```diff
@@ -10,6 +10,7 @@
 import struct
 import termios
 import time
+from nle.nethack.actions import _ACTIONS_DICT
```
Comment (heiner) [inline nle/scripts/ttyplay.py:13]: I don't like this as it makes ttyplay.py nethack-only all of a sudden. Better to add another script and import the relevant functions.

File: nle/tests/test_envs.py [hunk 10]
```diff
@@ -365,7 +365,7 @@ def test_final_reward(self, env):
 
         # Hack to quit.
         env.env.step(nethack.M("q"))
-        _, reward, done, _ = env.step(env._actions.index(ord("y")))
+        obs, reward, done, _ = env.step(env._actions.index(ord("y")))
```
Comment (heiner) [inline nle/tests/test_envs.py:368]: Why this change?

**Review by heiner — COMMENTED** [review]
This polybeast version here truly is a beast, in the bad sense of the word. I'm not convinced hydra is adding anything here either, but I understand this is just a copy&paste from the agent repo.

I can see how this might not be the easiest setup to start experimenting.

File: nle/env/base.py (hunk 6 above)
Comment (condnsdmatters) [reply nle/env/base.py:407]: This checks if death the game is over. will improve clarity

File: nle/scripts/play.py (hunk 8 above)
Comment (condnsdmatters) [reply nle/scripts/play.py:122]: if you are _not_ printing all frames, then you are going back and overwriting instead.  i wanted a `store_true` parameter that would stop the overwriting behaviour. howabout `print_frames_separately`?

File: nle/scripts/ttyplay.py (hunk 9 above)
Comment (condnsdmatters) [reply nle/scripts/ttyplay.py:13]: do you mean `nle` only?

Comment (condnsdmatters) [reply nle/scripts/ttyplay.py:13]: do you mean now you cant just take the script and play any old ttyrec, away from nle?

File: nle/tests/test_envs.py (hunk 10 above)
Comment (condnsdmatters) [reply nle/tests/test_envs.py:368]: i was undoing a previous change to the tests a few commits back, and slipped.

Comment (condnsdmatters) [reply nle/tests/test_envs.py:368]:
```suggestion
        _, reward, done, _ = env.step(env._actions.index(ord("y")))
```

**Comment by condnsdmatters** [comment]
I have rebased and restructured into 3 commits. I'm increasingly tempted to cut out the agent and move it to a branch - am happy to do this if we think this is sensible.

**Comment by condnsdmatters** [comment]
All tests passed so rebased to include latest changes.

**Comment by heiner** [comment]
> I have rebased and restructured into 3 commits. I'm increasingly tempted to cut out the agent and move it to a branch - am happy to do this if we think this is sensible.

I think moving the agent to a branch is a good idea.

**Comment by condnsdmatters** [comment]
The agent code now lives at #163. When this is merged, I will rebase that one off merge so the versions all make sense (currently the "challenge" environment doesnt exist in that branch.

**Comment by condnsdmatters** [comment]
(Also a quick drive by fix for the observation space bug spotted on Friday)

File: nle/nethack/actions.py [hunk 11]
```diff
@@ -76,6 +89,12 @@ class MiscAction(enum.IntEnum):
...
+
+
 class Command(enum.IntEnum):
```
Comment (heiner) [inline nle/nethack/actions.py:98]: Previously, this was "all commands" (or something), and the "RL relevant ones" we gathered in their own enum. What's the philosophy now?

File: nle/scripts/play.py [hunk 12]
```diff
@@ -67,7 +68,18 @@ def get_action(env, action_mode, is_raw_env):
...
+    env,
+    mode,
+    ngames,
+    max_steps,
+    seeds,
+    savedir,
+    no_render,
+    render_mode,
+    print_frames_separately,
+    **kwargs,
```
Comment (heiner) [inline nle/scripts/play.py:81]: What do we use the **kwargs for?

File: nle/scripts/ttyplay.py [hunk 13]
```diff
@@ -126,6 +135,12 @@ def process(f):
             continue
 
         if channel == 1:  # Input channel.
+            os.write(
+                1, b"\033[s\033[26;0f\033[37;1mFrame %d:\033[0m " % frame
+            )  # Save Cursor & Jump to L26
```
Comment (heiner) [inline nle/scripts/ttyplay.py:140]: could this interfere with other saved cursors from the data stream?

**Review by heiner — APPROVED** [review]
Generally looks good to me.

File: nle/scripts/ttyplay.py (hunk 13 above)
Comment (condnsdmatters) [reply nle/scripts/ttyplay.py:140]: A quick google suggests, yes you can only save one cursor at a time, so yes this would interfere. On test rendering i've not seen any problems though for our ttyrec2s...

File: nle/scripts/play.py (hunk 12 above)
Comment (condnsdmatters) [reply nle/scripts/play.py:81]: the `debug` flag is created but does nothing in this function because it has been used previously. however this fn is called with `play(**vars(flags))`, so it needs a home here. the home i thought would be clearer in `**kwargs`, but i almost put `**_`

File: nle/nethack/actions.py (hunk 11 above)
Comment (condnsdmatters) [reply nle/nethack/actions.py:98]: This enum wasn't really all commands because it didnt contain movement etc?  I figured if we were splitting out lets at least clearly mark the very dangerous ones.
